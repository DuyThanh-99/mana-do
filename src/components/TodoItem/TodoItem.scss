.todoitem {
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-bottom: 1px solid rgb(0 0 0 / 20%);
  padding-bottom: 5px;
  margin-bottom: 5px;
  span {
    flex: 1 1;
    text-align: left;
    margin-left: 8px;
  }
  &__delete {
    outline: none;
    border: none;
    width: 32px;
    height: 32px;
    min-width: auto;
    min-height: auto;
    box-shadow: none;
    border-radius: 50%;
    display: flex;
    justify-content: center;
    align-items: center;
    cursor: pointer;
  }
  &__option {
    width: 24px;
    height: 24px;
    cursor: pointer;
    background: linear-gradient(
      16.76deg,
      rgba(255, 255, 255, 0.9) 7.93%,
      rgba(196, 196, 196, 0.9) 81.39%
    );
    border-radius: 50%;

    &--checked {
      background: linear-gradient(136.67deg, #395ef2 10.93%, #67d1fa 89.03%);
      border: none;
      position: relative;
      &::after {
        content: "";
        position: absolute;
        left: 9px;
        top: 5px;
        width: 5px;
        height: 10px;
        border: solid #ffffff;
        border-color: #ffffff;
        border-width: 0px 3px 3px 0;
        -webkit-transform: rotate(40deg);
        -ms-transform: rotate(40deg);
        transform: rotate(40deg);
      }
    }
  }
  &__input {
    margin-bottom: 5px;
  }
}
