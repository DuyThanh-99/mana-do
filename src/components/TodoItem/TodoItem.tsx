import React, { Fragment, useEffect, useRef, useState } from "react";
import { MdOutlineDelete } from "react-icons/md";
import "./TodoItem.scss";

export const TodoItem = React.forwardRef<any, ITodoItem>(
  (
    {
      id,
      content,
      checked,
      onClick,
      onChange,
      idClick,
      setIdClick,
      handleEditTodos,
    },
    ref
  ) => {
    const isEditing = id === idClick;
    const editInputRef = useRef<HTMLInputElement>(null);
    const [text, setText] = useState<string>(content);
    /** Handle Click Outside */
    useEffect(() => {
      if (idClick) {
        const handleClickOutside = (event: any) => {
          if (
            editInputRef.current &&
            !editInputRef.current.contains(event.target)
          ) {
            setIdClick("");
            setText(content);
          }
        };
        document.addEventListener("click", handleClickOutside);
        return () => {
          document.removeEventListener("click", handleClickOutside);
        };
      }
    }, [ref, idClick, setIdClick, content]);

    /** Handle enter press key to edit */
    const handleOnKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
      if (e.key === "Enter" && text) {
        return handleEditTodos({ id: idClick, content: text });
      }
    };

    return (
      <Fragment>
        {!isEditing ? (
          <div className="todoitem">
            <input
              id={id}
              style={{ display: "none" }}
              type="checkbox"
              checked={checked}
              onChange={onChange}
              data-testid="TodoItem_checkbox"
            />
            <label htmlFor={id} data-testid="TodoItem_checkbox-label">
              <div
                className={`todoitem__option ${
                  checked ? "todoitem__option--checked" : ""
                }`}
              ></div>
            </label>
            <span
              onDoubleClick={() => {
                setIdClick(id);
                editInputRef.current?.focus();
              }}
              style={{
                textDecorationLine: checked ? "line-through" : "none",
                textDecorationColor: "grey",
              }}
              data-testid="TodoItem_content"
            >
              {content}
            </span>
            <MdOutlineDelete
              className="todoitem__delete"
              onClick={onClick}
              data-testid="TodoItem_btnDel"
            />
          </div>
        ) : (
          <input
            ref={editInputRef}
            className="todoitem__input"
            value={text}
            onChange={(e) => {
              setText(e.target.value);
            }}
            onKeyDown={handleOnKeyDown}
            data-testid="TodoItem_input"
          />
        )}
      </Fragment>
    );
  }
);
