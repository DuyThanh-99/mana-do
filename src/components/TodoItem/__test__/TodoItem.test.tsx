import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { TodoItem } from "../TodoItem";

type RenderComponentTypes = Pick<ITodoItem, "id" | "idClick">;

const onUpdate = jest.fn();
const onDelete = jest.fn();

const renderComponent = (props: RenderComponentTypes) => {
  const { id, idClick } = props;
  return render(
    <TodoItem
      id={id}
      idClick={idClick}
      handleEditTodos={onUpdate}
      content="OK"
      checked={false}
      onChange={jest.fn()}
      onClick={onDelete}
    />
  );
};

describe("<TodoItem/> edit", () => {
  it("should match snapshot", () => {
    const wrapper = renderComponent({ id: "1", idClick: "1" });
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should render UI correctly", () => {
    renderComponent({ id: "1", idClick: "1" });
    const input = screen.getByTestId("TodoItem_input");
    expect(input).toBeInTheDocument();
    expect(input).toHaveValue("OK");
  });

  it("should be able to type in input", () => {
    renderComponent({ id: "1", idClick: "1" });
    const input = screen.getByTestId("TodoItem_input") as HTMLInputElement;
    fireEvent.change(input, { target: { value: "Reading some books" } });
    expect(input.value).toBe("Reading some books");
  });
});

describe("<TodoItem/>", () => {
  it("should match snapshot", () => {
    const wrapper = renderComponent({ id: "1", idClick: "2" });
    expect(wrapper.container).toMatchSnapshot();
  });

  it("should check the checkbox Element is hidden", () => {
    renderComponent({ id: "1", idClick: "2" });
    const checkbox = screen.getByTestId(
      "TodoItem_checkbox"
    ) as HTMLInputElement;
    expect(checkbox).toHaveStyle(`display: none;`);
  });

  it("should call del func", () => {
    renderComponent({ id: "1", idClick: "2" });

    const btn = screen.getByTestId("TodoItem_btnDel");
    userEvent.click(btn);
    expect(onDelete).toBeCalledTimes(1);
  });
});
