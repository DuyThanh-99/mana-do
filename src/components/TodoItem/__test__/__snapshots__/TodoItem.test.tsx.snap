// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem/> edit should match snapshot 1`] = `
<div>
  <input
    class="todoitem__input"
    data-testid="TodoItem_input"
    value="OK"
  />
</div>
`;

exports[`<TodoItem/> should match snapshot 1`] = `
<div>
  <div
    class="todoitem"
  >
    <input
      data-testid="TodoItem_checkbox"
      id="1"
      style="display: none;"
      type="checkbox"
    />
    <label
      data-testid="TodoItem_checkbox-label"
      for="1"
    >
      <div
        class="todoitem__option "
      />
    </label>
    <span
      data-testid="TodoItem_content"
      style="text-decoration-line: none; text-decoration-color: grey;"
    >
      OK
    </span>
    <svg
      class="todoitem__delete"
      data-testid="TodoItem_btnDel"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
      />
    </svg>
  </div>
</div>
`;
